Bump minimum dependency versions

FranciscoPombal opened the pull request:
It doesn't make sense to allow anything lower (our floor at the moment is what the Ubuntu 18.04 repository has).

Also,

Closes https://github.com/qbittorrent/qBittorrent/issues/13510

glassez commented on CMakeLists.txt:
```diff
@@ -12,11 +12,11 @@ project(qBittorrent
-set(requiredLibtorrentVersion 1.2.0)
-set(requiredZlibVersion 1.2.5.2)
+set(requiredOpenSSLVersion 1.1)
+set(requiredLibtorrentVersion 1.2.0...2.0)
```
What this notation means?

FranciscoPombal replied:
It means "from 1.2.0, inclusive, up to 2.0, inclusive".

FranciscoPombal replied:
> Doesn't `set(requiredLibtorrentVersion 1.2.0)` mean any version starting from 1.2?

Yes, but that doesn't accurately reflect what we support as accurately as `1.2.0...2.0` does. If 2.1 is released tomorrow, we want qBittorrent to fail to build with it automatically, until we manually bump the supported version.

glassez replied:
How about 2.0.1, 2.0.2, etc.?

FranciscoPombal replied:
I would expect specifying `2.0` means "any patch/tweak version, as long as major and minor are `2.0`"

zywo commented:
@FranciscoPombal that doesn't solve #13510

FranciscoPombal commented:
> @FranciscoPombal that doesn't solve #13510

I'll try harder to figure out the true cause then, and adjust this PR accordingly.

FranciscoPombal commented:
This can be merged right away, but the libtorrent change will only have a real effect once https://github.com/arvidn/libtorrent/pull/5221 is merged to the relevant branches.

glassez commented:
>This can be merged right away

Don't forget to apply the same changes to automake files.

FranciscoPombal commented:
Big oof, I just realized version ranges are only available in CMake >= 3.19... they already updated the docs pages with the 3.19 RC, but I thought I was looking at some earlier version.

So I guess the ranges will have to wait, but that's not too bad. The main purpose of this PR is simply about raising minimum versions anyway.

FranciscoPombal commented:
@glassez 

> Don't forget to apply the same changes to automake files.

Done. Christ, what a shitty build system. Such a noisy diff because of it. Also it forced me to search through the files to find all instances where versions should be changed... hope I didn't miss any.

glassez commented:
Personally, I don't mind these increases. But maybe some of this is still unavailable on some systems? Although both boost and zlib are quite old. What about OpenSSL?

Chocobo1 commented:
>Personally, I don't mind these increases. 

Just curious, why? just because they are ancient?
We don't deny using newer versions (assuming they are necessary and widely available) but when we do it only then the version bumps will happen, isn't that the case?

glassez commented:
I don't think we should raise the requirements just because the previous versions are very old. But in fact (if I'm not mistaken), none of the active developers/testers of the project use such old versions (even close to them), we don't even care about the project being built with them (otherwise we would have to have CI configuration with the minimum supported versions). What is the point of us claiming that they are still supported? What if someone really hopes for it, and gets some errors? Do you really want to be responsible for supporting this junk?
I don't insist on raising them, of course. It may not have any problems. We can just wait for some problem to manifest itself in order to increase the requirements for the corresponding library.

Chocobo1 commented:
>We can just wait for some problem to manifest itself in order to increase the requirements for the corresponding library.

I agree, at least that would fix a real problem, so until it happens I wouldn't touch the numbers.

FranciscoPombal commented:
@Chocobo1 @glassez 

> > We can just wait for some problem to manifest itself in order to increase the requirements for the corresponding library.
> 
> I agree, at least that would fix a real problem (not hypothetical ones) and I think this falls into the "necessary" case in my previous reply.
> So until it happens I wouldn't touch the numbers.

These newer versions are already widely available in all platforms. The oldest platform we support in practice is Ubuntu 18.04 due to the Qt 5.9 requirement. Theses version bumps reflect dependency versions available at least in that system.

Personally, I don't want to waste time with Windows/macOS users who report bugs with a self-compiled version that uses Boost 1.40 and/or OpenSSL 1.0, for instance. It is very likely that there will be some bug due to asio or something that was fixed in later versions. Or maybe some HTTPS tracker announces will fail due to the tracker using TLS 1.3 and them using OpenSSL 1.0. So we should bump the versions to prevent these kinds of problems and reports.

We are a small team, and can't afford to support such a wide spectrum of stuff. Of course, our CI should test the minimum versions also on Windows/macOS - this is something that, as discussed previously, should be addressed in the future.(*)

But CI only gets you so far. If someone needs to investigate a bug report more thoroughly from someone that compiled with Boost 1.40, OpenSSL 1.0 and zlib 1.2.5.2, are you going to bother installing all those crappy old versions? Or are you going to tell the user to try to reproduce the problem with a more recent stack? In either case, we would have saved everyone's time by just requiring newer versions in the first place. It is misleading to pretend that compiling with any of these very old versions of the dependencies will lead to any kind of good result, otherwise, we would be doing it ourselves.
IMO, dependency version support shouldn't be only about "the project technically compiles with versions this old". It must also be about what's convenient to us. So I disagree with waiting until the last possible moment to bump, I think we should be proactive.

I agree that bumping between patch versions would have been a not so great choice. But we are on the verge of a major (in terms of size) release, which is also a new series - 4.3.x. It's not a bad time to bump the requirements. We already did so for libtorrent, why not the rest?

(*)Windows/macOS may be exceptions to this... As you know, we are always racing to support newer Qt versions in these platforms due to the very important high DPI and other types of fixes that are only available in newer versions. In practice, it is very undesirable for anyone to compile with Qt 5.9 on Windows 10, for instance. Sometimes I really wonder if we should have split Qt requirements for Linux and Windows/macOS. I wouldn't like that, but it is something to consider.

glassez commented:
>We are a small team, and can't afford to support such a wide spectrum of stuff.
If someone needs to investigate a bug report more thoroughly from someone that compiled with Boost 1.40, OpenSSL 1.0 and zlib 1.2.5.2, are you going to bother installing all those crappy old versions? Or are you going to tell the user to try to reproduce the problem with a more recent stack? In either case, we would have saved everyone's time by just requiring newer versions in the first place.

This coincides with my reasons for not objecting to raising the minimum version requirements. But if someone from our small team has the desire to mess with it and believes that the game is worth the candle, then let them do it. I don't have the ability to catch errors that occur in builds other than my own.
Another reason is our previous experience. I have often seen a situation where we even delayed a release while waiting for a new version of libtorrent that fixed some bug or made important improvements, but we still continued to support a bunch of previous versions. The question is why, if these latest changes are so important to us?

jagannatharjun commented:
> Another reason is our previous experience. I have often seen a situation where we even delayed a release while waiting for a new version of libtorrent that fixed some bug or made important improvements, but we still continued to support a bunch of previous versions. The question is why, if these latest changes are so important to us?

regular versions bump on libtorrent is kinda ok, but I really don't like to update zlib or openssl or boost, these are used in my other projects and I really don't want to break those for no particular reason. I will say let someone open some can of worms and if nothing else can be done then we can drop the support for that particular problem.

glassez commented:
By the way, as for the Boost... It doesn't seem to be a direct dependency, does it? Why not just reflect the corresponding requirements of libtorrent?

FranciscoPombal commented:
@glassez 

>But if someone from our small team has the desire to mess with it and believes that the game is worth the candle, then let them do it.

Someone on this team is actively testing with such older versions? Can you name one such person that has been doing this, or is doing this right now, or will do this in the future?

> I don't have the ability to catch errors that occur in builds other than my own.

Same. Sometimes a may go a bit out of my way for a slightly different setup, but not something totally different and outdated (like OpenSSL 1.0). Let's not forget that in some cases, like OpenSSL 1.0, it is even no longer supported upstream.

>Another reason is our previous experience. I have often seen a situation where we even delayed a release while waiting for a new version of libtorrent that fixed some bug or made important improvements, but we still continued to support a bunch of previous versions. The question is why, if these latest changes are so important to us?

A good point, and I agree with it. We need to be less fearful of bumping the minimum required versions for this reason. Early versions of libtorrent 1.2.x have too many very prominent bugs to be used with qBittorrent. Would you recommend anyone building qBittorrent on any platform with libtorrent 1.2.2, for instance? With such a version, on Windows, they will suffer from the auto cache bug, then there is also the tracker announce queue bug leading to memory leaks, etc, etc, etc (just look at the libtorrent changelog). No one in their right mind is compiling qBittorrent from source with libtorrent 1.2.x releases older than the latest RC_1_2 or 1.2.10, despite the fact that we technically allow >= 1.2.0. They would produce unnecessarily buggy builds, basically unusable.

@jagannatharjun 

> regular versions bump on libtorrent is kinda ok, but I really don't like to update zlib or openssl or boost, these are used in my other projects and I really don't want to break those for no particular reason.

If your other projects have problems with building with, say, non-EOL versions of OpenSSL, more than likely you've got bigger problems and are accruing severe technical debt. Assuming of course, these are not "dead" projects that you want to keep in a certain state. But in either case, you should consider somehow better isolating your build environments, honestly.

> I will say let someone open some can of worms and if nothing else can be done then we can drop the support for that particular problem.

I already explained why I don't think being so passive is a good idea. It's a waste of everyone's time.

Me and others (like @xavier2k6) have been building with bleeding edge versions for months. On my part, I recall having used Boost 1.71 through 1.74 so far (both Linux and Windows), and I also went through a few OpenSLL versions. Been using zlib 1.2.11 since I remember. Even Qt >= 5.15 doesn't support anything older than OpenSLL 1.1.1. We must keep up.

FranciscoPombal commented:
> By the way, as for the Boost... It doesn't seem to be a direct dependency, does it? Why not just reflect the corresponding requirements of libtorrent?

It's a direct dependency at lest due to the circular buffer, right?

glassez commented:
> > By the way, as for the Boost... It doesn't seem to be a direct dependency, does it? Why not just reflect the corresponding requirements of libtorrent?
> 
> It's a direct dependency at least due to the circular buffer, right?

Oh, that's right.

glassez commented:
> > But if someone from our small team has the desire to mess with it and believes that the game is worth the candle, then let them do it.
> 
> Someone on this team is actively testing with such older versions? Can you name one such person that has been doing this, or is doing this right now, or will do this in the future?

I only referred to a previous comment of @Chocobo1 where he called it a good idea to wait for one of these dependencies to cause an error in order to increase the requirements for the corresponding library. I assume that he is willing to handle such problems, trying to figure out exactly where they come from (from which libraries and which versions).

FranciscoPombal commented:
@glassez 

> I only referred to a previous comment of @Chocobo1 where he called it a good idea to wait for one of these dependencies to cause an error in order to increase the requirements for the corresponding library. 

Well, we both (https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-709981611) agree that this approach is too passive. The only permissible scenario where it would be acceptable would be if someone would be willing to invest the time and effort to play with such old stacks. Which brings us to...

> I assume that he is willing to handle such problems, trying to figure out exactly where they come from (from which libraries and which versions).

I can't speak for them, but is this really the case? @Chocobo1 are you actually willing to do this? Are you/Will you become the "designated solver of bug reports using very old dependency versions"?

Not to re-ignite this other debate, but some food for thought: the same reasoning applied here to tighten up version requirements also applies to OS support. \*\*cough\*\* Windows 7 support \*\*cough\*\*. @Chocobo1 do you have a Windows 7 VM where you test qBittorrent when a bug report about it comes in?

I only ask these questions because I have not seen this happen. But if there is a new commitment for the future, I can understand. Otherwise, it's a waste of time for everyone else.

glassez commented:
>the same reasoning applied here to tighten up version requirements also applies to OS support.

I do not think so. At least, if we are talking about Windows. Many programs perfectly support different versions of it, including quite old ones, without any tedious worries. If the program does not directly use some low-level API, then there should be no problems. Quite a large number of people still use Windows 7 (and I have serious assumptions that a lot of our users do too).

xavier2k6 commented:
**IF** minimum versions are going to be raised........I'd say for now, keep to these guidelines for libtorrent

### RC_1_1

>You'll need at least version **1.49** of the boost library in order to build libtorrent.

https://github.com/arvidn/libtorrent/blob/RC_1_1/docs/building.rst#building-with-bbv2

### RC_1_2

>You'll need at least version **1.58** of the boost library in order to build libtorrent.

https://github.com/arvidn/libtorrent/blob/RC_1_2/docs/building.rst

### RC_2_0

>You'll need at least version **1.66** of the boost library in order to build libtorrent.

https://github.com/arvidn/libtorrent/blob/RC_2_0/docs/building.rst

Although - just as an example:
`Boost 1.73`, more specifically `asio 1.16.1` fixed one issue #11039 & _may_ have fixed another #8740

FranciscoPombal commented:
@glassez 

> I do not think so. At least, if we are talking about Windows. Many programs perfectly support different versions of it, including quite old ones, without any tedious worries. If the program does not directly use some low-level API, then there should be no problems.

I understand your point about low-level APIs. However, nobody on the team uses Windows 7. There may be legitimate-looking reports that are in fact specific issues to Windows 7. Thus, again, for the same reasons as above, this is a waste of time (especially for people like me who are on top of the issue tracker all the time).

Besides, as you might recall, we are already accruing technical debt due to keeping Windows 7 support - there are problems compiling with modern MinGW versions because of it.

"just because it still technically compiles and appears to run fine, doesn't mean we are obliged to support it officially".

>Quite a large number of people still use Windows 7 (and I have serious assumptions that a lot of our users do too).

I have serious doubts about that. But that doesn't matter. If they are running Windows 7 still, they shouldn't, and we shouldn't be complicit in enabling that. It's insecure.

Contrary to what you may believe, the vast majority of the most vocal people who still use Windows 7 today do not do so by true necessity, but due to a toxic mix of laziness and "beginner-expert" tech illiteracy. These are the same kind of folks that will try to "debloat" Windows 10 and use stuff like "Windows 10 Ameliorated" (genius idea to trust an OS ISO from some rando online) instead of the real solution, which is moving to Linux.

Would there be backlash from this vocal segment of the userbase which would be unpleasant to deal with? Probably. But fuck them, if they don't want to listen to reason. We're not here to indulge their bad habits.

Now the economic argument: Windows 10 is free nowadays. You only lose out on minor customization/"Microsoft account" stuff if you don't pay for a license. Not to mentioned that the "accessibility upgrade loophole" was never closed, so it is _still_ possible to get a legitimate license that way, _even if starting from an illegitimate copy of Windows 7/8.1_.

If these people are using computers so old they can't even run Windows 10 reasonably, I doubt they're doing anything on a regular basis for which there isn't an adequate Linux replacement already, or for which WINE can't do a good job of running. CAD software, Photoshop et. al. is the kind of software that typically either doesn't run adequately in WINE or for which there isn't an adequate replacement, but are people running that kind of software in computers that can't even run Windows 10 properly?

xavier2k6 commented:
Again, **IF** minimum versions are going to be raised:

Should **Qt** be raised?

**Qt 5.9 LTS** => **Qt 5._12_ LTS**

I very rarely use **linux** & don't use **macOS** so I don't know if there's a "cross-platform" reason or whatever to keep it at **Qt 5.9 LTS** or whether the reason was to wait for the next **LTS** version before moving to it.....if that was the case then i don't see the reason not to go to **Qt 5._12_ LTS** as minimum.

FranciscoPombal commented:
> Again, **IF** minimum versions are going to be raised:
> 
> Should **Qt** be raised?
> 
> **Qt 5.9 LTS** => **Qt 5._12_ LTS**
> 
> I very rarely use **linux** & don't use **macOS** so I don't know if there's a "cross-platform" reason or whatever to keep it at **Qt 5.9 LTS** or whether the reason was to wait for the next **LTS** version before moving to it.....if that was the case then i don't see the reason not to go to **Qt 5._12_ LTS** as minimum.

I don't think so, because that means dropping out-of-the-box support for Ubuntu 18.04, which is still a well-supported release that many Linux users still use (including me). However, in practice, as mentioned above, no one is compiling with such old Qt versions on Windows/macOS. An (ugly) solution to that problem is having split version requirements for Qt depending on the OS as I mentioned at the bottom of https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-709972183. I think 18.04 support should be maintained _at least_ until 20.04.2 or 20.04.3 are released. Either way, in case anyone is thinking about this, it would be extremely idiotic and inconsistent to do something like dropping support for 18.04 before Windows 7.

xavier2k6 commented:
@FranciscoPombal **"Drop Windows 7 Support"** should be it's **own issue** I think as it's not fully relevant to this **PR**

(I don't know if it will be dropped before the current timeframe for next release 4.3.x unless it gets the backing of the maintainer/majority of team)

To aid in your arguments you can use the "Builds" of Windows 7 `7600` `7601` in the issue tracker search filters to show the significance of reports etc.

[7600](https://github.com/qbittorrent/qBittorrent/issues?q=is%3Aissue+is%3Aopen+7600)

[7601](https://github.com/qbittorrent/qBittorrent/issues?q=is%3Aissue+is%3Aopen+7601)

or even go through OS:Windows label

[OS:Windows](https://github.com/qbittorrent/qBittorrent/issues?q=is%3Aopen+label%3A%22OS:+Windows%22+is%3Aissue)

glassez commented:
> @FranciscoPombal **"Drop Windows 7 Support"** should be it's **own issue** I think as it's not fully relevant to this **PR**

👍

FranciscoPombal commented:
> > @FranciscoPombal **"Drop Windows 7 Support"** should be it's **own issue** I think as it's not fully relevant to this **PR**
> 
> +1

Like I said, I don't want to center discussion around it (it is quite enough at this point, I suppose), but it is certainly relevant. I'll create a standalone issue for it later.

Chocobo1 commented:
>I assume that he is willing to handle such problems, trying to figure out exactly where they come from (from which libraries and which versions).

Do such problems really exist in our issue tracker? I don't really recall we had such issues for a very long time, lets be specific, this is about: boost, openssl, zlib. If such issue arise in the future I'm willing to take a look (assuming I have enough free time) or even changing my mind on this topic.

Another idea to circumvent this topic is to drop the idea of "minimum **supported** version" and adopt another concept: "minimum **tested** version". This way the new version numbers don't reflect the code requirements and just shows what we are able to test (and therefore support). I'm OK with this idea, but we need to be clear about this in INSTALL file, cmake scripts and other places. For example, "version requirements" would need to be changed to "tested versions" and "requiredBoostVersion" to "testedBoostVersion", and/or maybe just a simple line explaining that this is the version number tested, not the real code requirement.
What do you guys think? Then we will really need to test the listed versions, I suppose this won't be a problem once the CI is configured to build against them.

xavier2k6 commented:
>this is about: boost, openssl, zlib

Ref Boost (libtorrent)

https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-710060113

This is probably the one dependency (Boost) that should be changed.....

FranciscoPombal commented:
@Chocobo1 

> Do such problems really exist in our issue tracker? I don't really recall we had such issues for a very long time, lets be specific, this is about: boost, openssl, zlib.

There are at least two examples mentioned here https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-710060113, concerning boost.

>  If such issue arise in the future I'm willing to take a look (assuming I have enough free time) or even changing my mind on this topic.

Why bother? Even libtorrent has much higher version requirements (it's at least 1.58 for 1.2.x). Leave the onus of building with something sane to the users. There are many users for very few of us. I think it's bad policy to commit to such a wide range of configurations with such little time and personnel. You're not even sure if you'll have the time for it, don't force yourself to do something that takes a lot of your time when each user could take just a little bit more of theirs to avoid the situation altogether.

> Another idea to circumvent this topic is to drop the idea of "minimum **required** version" and adopt another concept: "minimum **tested** version". This way the new version numbers don't reflect the code requirements and just shows what we are able to test (and therefore support). I'm OK with this idea, but we need to be clear about this in INSTALL file, cmake scripts and other places. For example, "version requirements" would need to be changed to "tested versions" and "requiredBoostVersion" to "testedBoostVersion", and/or maybe just a simple line explaining that these are merely the version numbers tested, not the real code requirement (old version might still work but you are on your own, etc).
What do you guys think? Then we will really need to test the listed versions, I suppose this won't be a problem once the CI is configured to build against them.

This is the spirit, it's basically what I've been trying to say (was it that unclear? if so, sorry about that). We just need to add a note to the README/INSTALL/build documentation files saying something like this:

```
You may be able to build with older versions of the dependencies, but support for such builds is not provided - you are on your own.
Ensure you are building with an officially supported configuration when reporting bugs.
Version requirements are bumped every once in a while, within reason, to keep the range of supported configurations manageable and get everyone the latest incremental improvements stemming from dependency upgrades at a reasonable pace, or exceptionally out of critical necessity.
You can expect a bumping pace roughly tied to what is required to support the latest 2 active Ubuntu LTS releases.
```

I also agree that, with this policy, `requiredXXXVersion` is probably not the best choice for those variable names, as it implies a hard minimum requirement.
Though I would go for `minSupportedXXXVersion` (or even `minXXXVersion` ) instead of `testedXXXVersion`, the latter which kind of implies the tested version is the only "blessed" one. The CMake scripts could have a simple comment near the versions reading `# Older versions may work, but you are on your own`.

glassez commented:
I would insist on two stricter rules:
1. Do not support versions of libraries that are not supported by our other dependencies. So it's silly for us to support a Boost older than libtorrent supports.
2. Stop supporting old libtorrent sub-versions as soon as the new one has important changes (for example, fixes a bug that affects qBittorrent, or adds a feature that we really need). It sounds unacceptable for me when the user is allowed to make a build that works obviously worse than the reference one.

NotTsunami commented:
> > Another reason is our previous experience. I have often seen a situation where we even delayed a release while waiting for a new version of libtorrent that fixed some bug or made important improvements, but we still continued to support a bunch of previous versions. The question is why, if these latest changes are so important to us?
> 
> regular versions bump on libtorrent is kinda ok, but I really don't like to update zlib or openssl or boost, these are used in my other projects and I really don't want to break those for no particular reason. I will say let someone open some can of worms and if nothing else can be done then we can drop the support for that particular problem.

zlib 1.2.11 is shipped with Ubuntu 18.04 if I recall correctly. What version/OS are you using out of curiosity?

FranciscoPombal commented:
@glassez 

> I would insist on two stricter rules:
> 
>     1. Do not support versions of libraries that are not supported by our other dependencies. So it's silly for us to support a Boost older than libtorrent supports.
> 
>     2. Stop supporting old libtorrent sub-versions as soon as the new one has important changes (for example, fixes a bug that affects qBittorrent, or adds a feature that we really need). It sounds unacceptable for me when the user is allowed to make a build that works obviously worse than the reference one.

Agreed. I would also add a 3rd:

3. (try hard to) drop support for EOL versions of dependencies as soon as possible (e.g. OpenSSL 1.0). In practice, if we're following rule 1 as we should, we should rarely have to do this "by hand" though.

@sledgehammer999  @Chocobo1 if you agree and there are no significant objections from others, I can force-push this PR to also bump the libtorrent requirement to 1.2.10 (or even 1.2.11 as it too [contains important tracker-related fixes](https://github.com/arvidn/libtorrent/blob/RC_1_2/ChangeLog), if @arvidn releases it soon).

Chocobo1 commented:
> > Do such problems really exist in our issue tracker? I don't really recall we had such issues for a very long time, lets be specific, this is about: boost, openssl, zlib.
> 
> There are at least two examples mentioned here [#13511 (comment)](https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-710060113), concerning boost.

Under "minimum required version" viewpoint, transitive version requirements doesn't really affect us IMO, but whatever.

>We just need to add a note to the README/INSTALL/build documentation files saying something like this:

It is mostly OK. FYI, I made a few adjustment:
```
Version numbers are bumped every once in a while to keep the range of supported configurations manageable.
You may be able to build with older versions of the dependencies, but support for such builds is not provided - you are on your own.
Ensure you are building with an officially supported configuration when reporting bugs.
```
I would like to keep the reason of version bumps opaque and avoid giving the idea of it tied to some specific linux distro.

>Though I would go for minSupportedXXXVersion (or even minXXXVersion ) instead of testedXXXVersion,

`minXXXVersion` is good for me.

Now with the "minimum tested/supported version" in place, I've no opinion on what the numbers should be. Just make sure this PR is properly adjusted and don't forget to handle the docs as well.

FranciscoPombal commented:
@Chocobo1 

> Under "minimum required version" viewpoint, transitive version requirements doesn't really affect us IMO, but whatever.

Boost is not just a transitive dependency, we depend on it directly for the circular buffer used in the speed plot at least (https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-710002313).


> It is mostly OK. FYI, I made a few adjustment:
> 
> ```
> Version numbers are bumped every once in a while to keep the range of properly tested configurations manageable.
> You may be able to build with older versions of the dependencies, but support for such builds is not provided - you are on your own.
> Ensure you are building with an officially supported configuration when reporting bugs.
> ```
> 
> I would like to keep the reason of version bumps opaque and avoid giving the idea of it tied to some specific linux distro.

> `minXXXVersion` is good for me.

Good points, sounds reasonable, LGTM :+1:
 
> Now with the "minimum tested/supported version" in place, I've no opinion on what the numbers should be. Just make sure this PR is properly adjusted and don't forget to handle the docs as well.

Alright, I'll make the required changes. As for the numbers themselves, I would go with the current numbers, except for libtorrent which I would like to change - but before proceeding, I'd like your comment on https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-710199169 and also @sledgehammer999 's input if possible.

glassez commented:
>also bump the libtorrent requirement to 1.2.10 (or even 1.2.11 as it too contains important tracker-related fixes, if @arvidn releases it soon).

👍 
1.2.11 is also required for further changes.

xavier2k6 commented:
> > also bump the libtorrent requirement to 1.2.10 (or even 1.2.11 as it too contains important tracker-related fixes, if @arvidn releases it soon).
> 
> 👍
> 1.2.11 is also required for further changes.

Libtorrent 1.2.11 is **released** ref: https://github.com/arvidn/libtorrent/releases/tag/v1.2.11

>I don't think so, because that means dropping out-of-the-box support for Ubuntu 18.04

Doesn't Ubuntu 18.04 come with Qt 5.9.5? so that could potentially be raised from 5.9.0 to become the minimum supported version?

glassez commented:
>Libtorrent 1.2.11 is released ref

So we can raise libtorrent version requirement as we discussed before and clean the code according to it.

FranciscoPombal commented:
Alright, now that vcpkg has libtorrent 1.2.11, we can proceed with the version bumps across the board.

I addressed all comments (added the required documentation text, renamed the CMake variables in accordance with what was agreed to `minXXXVersion`, ...).

I rolled up the changes from https://github.com/qbittorrent/qBittorrent/pull/13783 into this one (I added @glassez as co-author), and bumped the vcpkg version.

This is now ready to merge (assuming the CI is successful, which I'll check back on later).

glassez commented:
>I rolled up the changes from #13783 into this one 

I don't like the way you did it.
First, #13783 is something more than just increasing the supported version number (than this PR was originally). It actually removes support for older versions from the code. So it's incorrect to just squash it here. At least it should be separate (original) commit.
Second, it is illogical to insert approved changes (#13783) as part of the still (at least formally) not approved ones, thereby delaying the merging of the former.

xavier2k6 commented:
>set(minBoostVersion 1.65)

Shouldn't that be `minBoostVersion`**1.66** so that libtorrent 2.x can be built without issues at compile time etc as per https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-710060113 otherwise this will have to be bumped again via another PR.

FranciscoPombal commented:
@glassez 

> I don't like the way you did it.
First, #13783 is something more than just increasing the supported version number (than this PR was originally). It actually removes support for older versions from the code. So it's incorrect to just squash it here.

"Bump minimum dependency versions" implies removing support in the code, no? I was going to make these changes anyway as well, I was just waiting on vcpkg.

> At least it should be separate (original) commit.

OK, that's not a problem.

>Second, it is illogical to insert approved changes (#13783) as part of the still (at least formally) not approved ones, thereby delaying the merging of the former.

What exactly is not approved about this? My reading of the these latest comments (https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-710209838 and https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-710220287) is that there was consensus with the changes thus far.

@xavier2k6 

> Doesn't Ubuntu 18.04 come with Qt 5.9.5? so that could potentially be raised from 5.9.0 to become the minimum supported version?

OK.

> Shouldn't that be `minBoostVersion`**1.66** so that libtorrent 2.x can be built without issues at compile time etc as per https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-710060113 otherwise this will have to be bumped again via another PR.

That would imply dropping support for Ubuntu 18.04, which I don't think anyone wants right now. We could look into adding some logic to the build system for requiring a different minimum boost version based on the libtorrent version used, or OS (since you mentioned in a previous comment that boost >= 1.73 fixes some issues, and it would be possible to require that for Windows/macOS).

glassez commented:
> "Bump minimum dependency versions" implies removing support in the code, no?

At least it should do.


> I was going to make these changes anyway as well, I was just waiting on vcpkg.

Unfortunately, you did not give any notice of this intention. What's more, it looks like you didn't even intend to actually do it. You explicitly declared this PR ready to merge without having these changes: https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-709350061.


> > Second, it is illogical to insert approved changes (#13783) as part of the still (at least formally) not approved ones, thereby delaying the merging of the former.
> 
> What exactly is not approved about this?

What does it matter? There is PR that is already approved, so you just need to click on the "Merge" button. Why should we delay this just so that it becomes part of another PR? What's the point? It doesn't exist for me. Even if, say, the raise of other versions is approved after only a couple of hours, we will simply merge it after.

xavier2k6 commented:
>That would imply dropping support for Ubuntu 18.04

Why is that? (I haven't looked in to it myself)

>We could look into adding some logic to the build system for requiring a different minimum boost version based on the libtorrent version used

Why not set one & forget it - use libtorrent 2.x as the `base` (This way we know that 1.2.x/2.x will work if they have same min version)?!

>or OS (since you mentioned in a previous comment that boost >= 1.73 fixes some issues, and it would be possible to require that for Windows/macOS).

Is there OS limitations for Boost versions??

Could also make the same point about raising libtorrent limits to as of now `1.2.11` that we know fix issues as for raising >= boost 1.73 but that's probably for another time?? can stay with libtorrent guidelines for now........

Unless it merits being discussed further before merging.

FranciscoPombal commented:
@glassez 

> Unfortunately, you did not give any notice of this intention. What's more, it looks like you didn't even intend to actually do it. You explicitly declared this PR ready to merge without having these changes: #13511 (comment).

That comment was written long ago, when there was no expectation of the libtorrent release being ready soon, so at the time, I thought that we might as well merge what was ready. After 1.2.11 was released, I thought it was understood that this PR could be updated to reflect that (it is my interpretation of something you said yourself: https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-728725987).

> What does it matter? There is PR that is already approved, so you just need to click on the "Merge" button. Why should we delay this just so that it becomes part of another PR? What's the point? It doesn't exist for me. Even if, say, the raise of other versions is approved after only a couple of hours, we will simply merge it after.

My point is that these changes are approved (just no "formally", with "Approve" button, but given @Chocobo1's last comments, I think the intention is clear), a fact that so far you have not disputed, contrary to what you said:

> Second, it is illogical to insert approved changes (#13783) as part of the still (at least formally) not approved ones,

If the changes in #13783 are also approved, then what's the point of splitting the work? If you have any further requested changes, please post them here.

btw, I have reported the vcpkg build failure: https://github.com/microsoft/vcpkg/issues/14941

FranciscoPombal commented:
@xavier2k6 

> Why is that? (I haven't looked in to it myself)

Ubuntu 18.04 does not natively support anything newer than 1.65.x: https://packages.ubuntu.com/bionic/libboost-all-dev

> Why not set one & forget it - use libtorrent 2.x as the `base` (This way we know that 1.2.x/2.x will work if they have same min version)?!

> Is there OS limitations for Boost versions??
> 
> Could also make the same point about raising libtorrent limits to as of now `1.2.11` that we know fix issues as for raising >= boost 1.73 but that's probably for another time?? can stay with libtorrent guidelines for now........
> 
> Unless it merits being discussed further before merging.

If we wish to easily support widely used Linux distributions with "LTS" releases such Ubuntu, Debian, and derivatives, which are the most widely used ones, we must stick to the Boost version available in their repositories. Same reasoning for Qt, and same for OpenSSL and zlib (though in the case of these latter 2 dependencies, the versions available in the repositories of the "LTS" distros that we care about are also the very latest ones).

FranciscoPombal commented:
Original CI failure has been solved, but there is another expected failure, from a problem that I've also already reported here: https://github.com/microsoft/vcpkg/issues/14943.

FranciscoPombal commented:
@glassez @Chocobo1 

CI should be fixed now.

glassez reviewed (approved):
Approved. I don't want to stick around any longer.

zero77 commented:
Don't know if it helps but, the following source is good for tracking supported builds across multiple distros.

https://repology.org/project/

Using qbittorrent as an example:
https://repology.org/project/qbittorrent/badges

xavier2k6 commented:
With Ubuntu 18.04 now dropped, we can bump Boost dependency to `1.66` as per https://github.com/qbittorrent/qBittorrent/pull/13511#issuecomment-710060113 to be inline with ongoing `libtorrent 2.x` implementation
